google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/SeekableByteChannelPrefetcher.java

```diff
@@ -0,0 +1,494 @@
+   *   <li> else, call close()
+   * </ol>
+   */
+  private static class WorkUnit implements Callable<ByteBuffer>, Closeable {
+    public final ByteBuffer buf;
+    public long blockIndex;
+    private final SeekableByteChannel chan;
+    private final int blockSize;
+    private Future<ByteBuffer> futureBuf;
+
+    public WorkUnit(SeekableByteChannel chan, int blockSize, long blockIndex) {
+      this.chan = chan;
+      this.buf = ByteBuffer.allocate(blockSize);
+      this.futureBuf = null;
+      this.blockSize = blockSize;
+      this.blockIndex = blockIndex;
+    }
+
+    @Override
+    public ByteBuffer call() throws IOException {
+      long pos = ((long)blockSize) * blockIndex;
+      if (pos > chan.size()) {
+        return null;
+      }
+      if (pos < 0) {
+        // This should never happen, if the code's correct.
+        throw new IllegalArgumentException("blockIndex " + blockIndex + " has position " + pos + ": negative position is not valid.");
+      }
+      chan.position(pos);
+      // read until buffer is full, or EOF
+      while (chan.read(buf) >= 0 && buf.hasRemaining()) {}
+      return buf;
+    }
+
+    public ByteBuffer getBuf() throws ExecutionException, InterruptedException {
+      return futureBuf.get();
+    }
+
+    public WorkUnit resetForIndex(long blockIndex) {
+      this.blockIndex = blockIndex;
+      buf.clear();
+      futureBuf = null;
+      return this;
+    }
+
+
+    @Override
+    public void close() throws IOException {
+      chan.close();
+    }
+  }
+
+  /**
+   * Wraps the provided SeekableByteChannel within a SeekableByteChannelPrefetcher,
+   * using the provided buffer size.
+   *
+   * @param bufSize buffer size in bytes
+   * @param chan channel to wrap in the prefetcher
+   */
+  private SeekableByteChannelPrefetcher(SeekableByteChannel chan, int bufSize) throws IOException {
+    Preconditions.checkArgument(!(chan instanceof SeekableByteChannelPrefetcher),"Cannot wrap a prefetcher with a prefetcher.");
+
+    if (!chan.isOpen()) {
+      throw new IllegalArgumentException("channel must be open");
+    }
+    this.chan = chan;
+    if (bufSize <= 0) {
+      throw new IllegalArgumentException("bufSize must be positive");
+    }
+    this.size = chan.size();
+    if (bufSize > this.size) {
+      this.bufSize = (int)this.size;
+    } else {
+      this.bufSize = bufSize;
+    }
+    this.open = true;
+    int prefetcherIndex = prefetcherCount++;
+    // Make sure the prefetching thread's name indicate what it is and
+    // which prefetcher it belongs to (for debugging purposes only, naturally).
+    String nameFormat = "nio-prefetcher-" + prefetcherIndex + "-thread-%d";
+    ThreadFactory threadFactory = new ThreadFactoryBuilder()
+        .setNameFormat(nameFormat)
+        .setDaemon(true)
+        .build();
+    // Single thread to ensure no concurrent access to chan.
+    exec = Executors.newFixedThreadPool(1, threadFactory);
+  }
```

Comment:
Here's an idea that could give you an edge on every other GCS client library, e.g. [com.google.cloud.hadoop.gcsio.GoogleCloudStorageReadChannel](https://github.com/GoogleCloudPlatform/bigdata-interop/blob/47b7988/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageReadChannel.java).

Imagine you had these data structures:

- Assume `FileOffset` is `Long` in the range [0..2**64)
- Assume `ChunkId` is `Long` which you auto-increment
- [`TreeRangeMap<FileOffset, Future<ByteBuffer>>`](https://google.github.io/guava/releases/16.0/api/docs/com/google/common/collect/RangeMap.html) stores pending reads
- `TreeRangeMap<FileOffset, ChunkId>` stores pointers for finished reads
- `HashMap<ChunkId, SoftReference<ByteBuffer>>` stores data for finished reads
  - ProTip: [`new MapMaker().softValues().makeMap()`](https://google.github.io/guava/releases/16.0/api/docs/com/google/common/collect/MapMaker.html#weakValues())

Thanks to [`SoftReference`](https://docs.oracle.com/javase/7/docs/api/java/lang/ref/SoftReference.html) that will magically turn all your read operations into one gigantic LRU cache. Please note the size of the cache is tuned by JVM flags like `-Xmx30G` and this could use a lot of RAM if you're using `SeekableByteChannel#position` to binary search lots of terabyte-sized GCS objects all at once. This is a reasonable tradeoff IMHO, since 1) it makes the API zero-config; 2) GCS iops cost literal money; and 3) they've got lots of latency.

Then you have a 1MB or so default chunk size. Then each read() and prefetch read() will call [`subRangeMap`](https://google.github.io/guava/releases/16.0/api/docs/com/google/common/collect/RangeMap.html#subRangeMap(com.google.common.collect.Range)) to see if it care fetch less than a megabyte. Please note you'll need to be smart here, in the case of fragmentation. You'll also need to capture hard references to what already exists in the range, as you perform the read.

Now you know why I didn't put seek() support in the original impl :)

**Note:** This assumes GCS architecture actually makes `Content-Range` requests efficient. I know some storage layers, e.g. SQLite, treat individual blobs like a linked-list data structure, where even if you seek, each op still pays an O(n) cost. At cursory examination, I was unfortunately not able to determine the complexity of these operations. But you can do it anyway for great justice and lulz.

Reply:
This code has actually be running on GATK for a while and we're happy with it. We have two options: we can try to optimize it at the cost of having to retest it thoroughly, or we can leave it mostly as is.

I would lean towards minimal changes myself. The Prefetcher is designed and optimized for sequential accesses, so clients will not use it if they have the sort of random access patterns that would benefit from the approach you are describing. The biggest cost of the proposed approach is that is requires extensive modification to this code, but I'm also worried about the extra GC load it would impose unnecessarily (in the current code the old buffers are unreachable and thus collected on every GC. With soft references they would likely be kept in the first GC (because of the recent usage), and then if memory is still scarce we'd have a second GC round to clear the buffers. The other downside is that we'd be constantly allocating new buffers instead of (as currently) reusing the same ones over and over again.

So while your suggestion is nice, I think it would best be implemented as a separate caching class, one that describes itself as more suitable for random access than this class. There is an awful lot of sequential accesses in practice and it's nice to have code that is specifically designed for this case.